Correctly check if index settings have been updated

ellnix opened the pull request:
# Pull Request

## Related issue
Fixes #280 (partly)

- In issue #280 there is also the related problem of a redundant request to create an already existing index the first time it is used. However this request should not be significant to performance, so we can deal with it in a better way after some refactoring.

## PR checklist
Please check if your PR fulfills the following requirements:
- [X] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [X] Have you read the contributing guidelines?
- [X] Have you made sure that the title is accurate and descriptive of the changes?

# Notes (please read)

I believe this fixes issue #280 (I could not reproduce it after this edit). I will add regression tests when I have the time (it took a while trying to figure out what was going wrong).

The problem was (to the extent I have determined it):
- The first time an index is used `ms_ensure_init` will fetch it from the MeiliSearch instance
- Subsequent calls for that same index will have been memoized, so this problem occurs only once (makes it harder to test), however in background jobs this memoization is not relevant
- During that first call when the index is first fetched from the MeiliSearch instance, its settings are checked with the current settings to see if they need updating, from here there are 2 separate issues:
  1. The previous settings are not fetched correctly: [this line](https://github.com/meilisearch/meilisearch-rails/blob/e7a12569f18ee40684c3c643a3cf573f3465c62e/lib/meilisearch-rails.rb#L744C37-L744C37) provides an argument to `MeiliSearch::Client#settings` which takes no arguments and therefore causes an `ArgumentError` which is then rescued with nil (moral of the story always specify the exception you are `rescue`-ing)
  ```ruby
  current_settings = @ms_indexes[MeiliSearch::Rails.active?][settings].settings(getVersion: 1) rescue nil
  ```
  2. The [`meilisearch_settings_changed?`](https://github.com/meilisearch/meilisearch-rails/blob/e7a12569f18ee40684c3c643a3cf573f3465c62e/lib/meilisearch-rails.rb#L797) method has therefore never been run up to this point (as far as I can see) and has several problems that have therefore never been addressed
     1. The `prev` keys are not just strings instead of symbols, they are also camelCase-d
     2. There is no guaranteed order to `current` and `prev` values if they are arrays, and `==` comparison between Arrays will be false if their elements have different orders

brunoocasali commented on lib/meilisearch-rails.rb:
```diff
@@ -797,15 +797,17 @@ def ms_pk(options = nil)
+          current_value.map!(&:to_s).sort! if current_value.is_a?(Array)
+          prev_value.map!(&:to_s).sort! if prev_value.is_a?(Array)
+
+          return true if current_value != prev_value
```
You can't use `return` in a block as far as I remember 😬

brunoocasali commented on lib/meilisearch-rails.rb:
```diff
@@ -741,7 +741,7 @@ def ms_ensure_init(options = nil, settings = nil, index_settings = nil)
         @ms_indexes[MeiliSearch::Rails.active?][settings] = SafeIndex.new(ms_index_uid(options), meilisearch_options[:raise_on_failure], meilisearch_options)
 
-        current_settings = @ms_indexes[MeiliSearch::Rails.active?][settings].settings(getVersion: 1) rescue nil # if the index doesn't exist
+        current_settings = @ms_indexes[MeiliSearch::Rails.active?][settings]&.settings
```
+1

brunoocasali reviewed (commented):
I would love to have a test to ensure this bug is fixed, and I'm willing to wait to merge this PR until we have at least a clear way to reproduce this (I know this is hard I've tried as well).

Unless we could at least ensure the people who reported the issue tried the solution in staging/production and say to us they don't get the error anymore.

ellnix commented:
> I would love to have a test to ensure this bug is fixed, and I'm willing to wait to merge this PR until we have at least a clear way to reproduce this (I know this is hard I've tried as well).

I'll get on that.

> Unless we could at least ensure the people who reported the issue tried the solution in staging/production and say to us they don't get the error anymore.

I tagged them on the original issue discussion a month ago but haven't heard from them. I was hoping they could save me the trouble of writing the test but oh well, I'll figure it out.

ellnix commented:
@brunoocasali You were right to ask for a test for this, most of the logic was incorrect. I believe I have fixed it now.

I also rebased it, so it should be ready to merge with your approval. I am hoping to change the implementation of this in a refactor.

brunoocasali commented on lib/meilisearch-rails.rb:
```diff
@@ -800,19 +805,22 @@ def ms_pk(options = nil)
+        user_configuration.any? do |key, user|
+          server = server_state[key]
+
+          if user.is_a?(Hash) && server.is_a?(Hash)
+            meilisearch_settings_changed?(server, user)
+          elsif user.is_a?(Array) && server.is_a?(Array)
+            user.map(&:to_s) != server.map(&:to_s)
+          else
+            user.to_s != server.to_s
```
Can we find a way to reduce the `.to_s` calls here?

brunoocasali reviewed (approved):
Just a comment, and then we're good to go!

brunoocasali commented:
Hey @ellnix, thanks a lot for your time here, it was an amazing piece of work, I will merge this because it will be very important to the users to have this fix :)

bors merge
